plutus-core/plutus-ir/src/PlutusIR/Subst.hs

```diff
@@ -1,99 +1,124 @@
 {-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE RankNTypes          #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 module PlutusIR.Subst
-    ( uniquesTerm
-    , uniquesType
-    , fvTerm
+    ( fvTerm
     , ftvTerm
     , fvBinding
     , ftvBinding
     , ftvTy
+    , vTerm
+    , tvTerm
+    , tvTy
     ) where
 
 import PlutusCore.Core.Type qualified as PLC
-import PlutusCore.Name qualified as PLC
-import PlutusCore.Subst (ftvTy, uniquesType)
+import PlutusCore.Subst (ftvTy, ftvTyCtx, tvTy)
 
-import PlutusIR.Compiler.Datatype
 import PlutusIR.Core
 
 import Control.Lens
+import Control.Lens.Unsound qualified as Unsound
 import Data.Set as S hiding (foldr)
 import Data.Set.Lens (setOf)
-
-uniquesTerm
-    :: PLC.HasUniques (Term tyname name uni fun ann)
-    => Term tyname name uni fun ann -> Set PLC.Unique
-uniquesTerm = setOf termUniquesDeep
+import Data.Traversable (mapAccumL)
+import PlutusCore.Core.Plated (typeTyVars)
 
 -- | Get all the free term variables in a PIR term.
-fvTerm :: Ord name => Term tyname name uni fun ann -> Set name
-fvTerm = \case
-    Let _ NonRec bs tIn ->
-        let fvLinearScope b acc = fvBinding b
-                                   <> (acc \\ setOf bindingNames b)
-        in foldr fvLinearScope (fvTerm tIn) bs
-
-    Let _ Rec bs tIn ->
-        (foldMap fvBinding bs <> fvTerm tIn)
-        \\ setOf (traversed.bindingNames) bs
-
-    LamAbs _ n _ t -> delete n $ fvTerm t
-    Apply _ t1 t2 -> fvTerm t1 <> fvTerm t2
-    Var _ n -> singleton n
-    TyAbs _ _ _ t ->   fvTerm t
-    TyInst _ t _ ->    fvTerm t
-    Unwrap _ t ->      fvTerm t
-    IWrap _ _ _ t ->   fvTerm t
-    Constant{}       -> mempty
-    Builtin{}        -> mempty
-    Error{}          -> mempty
+fvTerm :: Ord name => Traversal' (Term tyname name uni fun ann) name
+fvTerm = fvTermCtx mempty
+
+fvTermCtx :: Ord name => S.Set name -> Traversal' (Term tyname name uni fun ann) name
+fvTermCtx bound f = \case
+    Let a r@NonRec bs tIn ->
+        let fvLinearScope bound' b = (bound' `union` setOf bindingNames b, fvBindingCtx bound' f b)
+            (bound'', bs') = mapAccumL fvLinearScope bound bs
```

Comment:
there must be some more elegant way to do this, but I cannot come up with sth

Reply:
Other than the `bound'` and `bound''` names, this looks fairly elegant to me :smile:

Reply:
Yeah, I also don't like it but it was the best I could do.

Reply:
I improved the names a bit.